packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -360,111 +400,340 @@ contract AmmPool is IBlockReceiver {
-                require(withdrawal.fee == 0, "INVALID_TX_DATA");
-                require(withdrawal.to == address(this), "INVALID_TX_DATA");
-                require(withdrawal.extraData.length == 0, "INVALID_TX_DATA");
-                // Now approve this withdrawal
-                withdrawal.validUntil = 0xffffffff;
-                bytes32 txHash = WithdrawTransaction.hashTx(exchangeDomainSeparator, withdrawal);
+        // Mint liquidity tokens
+        mint(join.owner, join.poolAmountOut);
+    }
+
+    function processExit(
+        Context  memory ctx,
+        PoolExit memory exit,
+        bytes    memory signature
+        )
+        internal
+    {
+        bytes32 poolTxHash = hashPoolExit(ctx.DOMAIN_SEPARATOR, exit);
+        if (signature.length == 0) {
+            require(queue[queuePos].txHash == poolTxHash, "NOT_APPROVED");
+            delete queue[queuePos];
+            queuePos++;
+        } else {
+            require(poolTxHash.verifySignature(exit.owner, signature), "INVALID_SIGNATURE");
+        }
+
+        uint poolTotal = totalSupply();
+        uint ratio = (exit.poolAmountIn * BASE) / poolTotal;
+
+        for (uint i = 0; i < ctx.tokens.length; i++) {
+            uint amount = ctx.ammBalances[i] * ratio / BASE;
+            require(amount >= exit.minAmountsOut[i], "LIMIT_OUT");
+            if (exit.toLayer2) {
+                TransferTransaction.Transfer memory transfer = ctx._block.readTransfer(ctx.txIdx++);
+                require(transfer.fromAccountID == accountID, "INVALID_TX_DATA");
+                require(transfer.from == address(this), "INVALID_TX_DATA");
+                require(transfer.to == exit.owner, "INVALID_TX_DATA");
+                require(transfer.tokenID == ctx.tokens[i].tokenID, "INVALID_TX_DATA");
+                require(isAlmostEqual(transfer.amount, amount), "INVALID_TX_DATA");
+                require(transfer.fee == 0, "INVALID_TX_DATA");
+                // Now approve this transfer
+                transfer.validUntil = 0xffffffff;
+                bytes32 txHash = TransferTransaction.hashTx(ctx.exchangeDomainSeparator, transfer);
                 exchange.approveTransaction(address(this), txHash);
-                numTransactionsConsumed++;
-                // Total balance in this contract increases by the amount withdrawn
-                totalBalance[tokens[i].addr] = totalBalance[tokens[i].addr].add(amount);
+                ctx.numTransactionsConsumed++;
+                // Update the balances in the account
+                ctx.ammBalancesInAccount[i] = ctx.ammBalancesInAccount[i].sub(amount);
+            } else {
+                // Make the amount available for withdrawing
+                balance[exit.owner][ctx.tokens[i].addr] = balance[exit.owner][ctx.tokens[i].addr].add(amount);
             }
+            ctx.ammBalances[i] = ctx.ammBalances[i].sub(amount);
+        }
+
+        // Burn liquidity tokens
+        burn(exit.owner, exit.poolAmountIn);
```

Comment:
I lock them up currently, but not that much of a problem now because they can't be transferred. But this is indeed the most important thing we HAVE to do onchain (other operations not really needed onchain). The user needs to prove he has N liquidity tokens so that he can withdraw from the pool.

> the operator's transaction will fail

This should be valid (but predictable). Otherwise the the pool can be manipulated shortly and the user may withdraw unexpected amounts.